packages/backend-common/src/context/types.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DateTime, Duration } from 'luxon';
+
+/**
+ * A function that accepts a context and produces a new, derived context from,
+ * decorated with some specific behavior.
+ *
+ * @public
+ */
+export type ContextDecorator = (ctx: Context) => Context;
+
+/**
+ * A context that is meant to be passed as a ctx variable down the call chain,
+ * to pass along scoped information and abort signals.
+ *
+ * @public
+ */
+export interface Context {
```

Comment:
I'm still a fan of the much more minimal `Context` interface in https://github.com/backstage/backstage/pull/5312/files#diff-89a2ef91b1efe5c84c025ee2db34aba2f91a665f2cc12856f70e915209a205f1

The trouble with a fat context interface is that we can't add or remove any required methods or parameters without a breaking change, making it a poor place to put utility methods.

For example, the current interface doesn't have a `withDeadline` method, and we couldn't add that without a breaking change. In that case an optional `withDeadline` method would be pretty much useless as well, as you would have to always be able to fall back to computing the delta instead. What would instead happen in practice is that me ship it as `Contexts.withDeadline`, `contextWithDeadline` or something like that, leaving people wondering why there are two different ways to access things.

I think the [Go Context](https://pkg.go.dev/context#Context) interface is pretty well proven, and translated to TypeScript it would just be something like

```ts
interface Context {
  readonly deadline?: Date // I think it's best to avoid luxon in core interfaces
  readonly abortSignal: AbortSignal;
  getValue<T>(key: unknown): T;
}
```

The only missing piece there is the reason the context was aborted, [which is actually being worked on](https://github.com/whatwg/dom/pull/1027)

Reply:
I know it's not a huge deal really, but one thing that grinds my gears a bit is the mutation patterns not being chainable. :) I think we mentioned that briefly when making the original context implementation.

```ts
callFunction(
  Contexts.withLabels(
    Contexts.withRequestData(
      Contexts.withBlah(ctx, a),
      request
    ),
    { label: 'value' })
  );
);
```

or alternatively

```ts
let derived = Contexts.withBlah(ctx, a);
derived = Contexts.withRequestData(derived, request);
derived = Contexts.withLabels(derived, { label: 'value' });
callFunction(ctx);
```

so would be nice to have something like the `with`.

```ts
callFunction(
  ctx.with(
    Contexts.withBlah(a),
    Contexts.withRequestData(request),
    Contexts.withLabels({ label: 'value' }),
  ),
);
```

I'd totally be up for condensing the context down as proposed, in terms of data, to the `deadline` + `abortSignal` + `getValue` though apart from that. I mean we could of course even remove `deadline` and `abortSignal` but I think golang does strike a good balance there in terms of having a really minimal essential set of things. And besides, the abort signal being in there, is also related directly to other things (it interacts with manual cancellation, timeouts, AND shutdown mechanisms) so having it "outside" the core while being effectlively 100% essential may not be helpful.

Reply:
Another way to get chainability if we feel it's a thing that we should have would be this

```ts
callFunction(
  Contexts.decorate(ctx).withLabels({ label: 'value' }).withRequestData(request).withBlah(a).get()
);
```

Alternatively even having the returned type of all of the `with*` methods be `Context & ContextDecorators`, then you could do

```ts
callFunction(
  Contexts.withLabels(ctx, { label: 'value' }).withRequestData(request).withBlah(a)
);
```

Another pattern:

```ts
callFunction(
  decorateContext(ctx).withLabels({ label: 'value' }).withRequestData(request).withBlah(a)
);
```

Or because I think chaining stuff end up being pretty meh in JS very often, tricky to test etc, perhaps just

```ts
callFunction(
  Contexts.wrap(ctx, {
    labels: { label: 'value' },
    requestData: request,
    blah: a,
  })
);
```

Biggest difference here is that it is much easier to evolve an interface that is only consumed, rather than one that is both produced and consumed by user code